collect_app/src/main/java/org/odk/collect/android/http/CollectServerClient.java

```diff
@@ -0,0 +1,157 @@
+                parser.setInput(isr);
+                parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, true);
+                doc.parse(parser);
+                isr.close();
+                isr = null;
+            } finally {
+                if (isr != null) {
+                    try {
+                        // ensure stream is consumed...
+                        final long count = 1024L;
+                        while (isr.skip(count) == count) {
+                            // skipping to the end of the http entity
+                        }
+
+                        isr.close();
+
+                    } catch (IOException e) {
+                        // no-op
+                        Timber.e(e, "Error closing input stream reader");
+                    }
+                }
+
+                if (is != null) {
+                    try {
+                        is.close();
+                    } catch (IOException e) {
+                        Timber.e(e, "Error closing inputstream");
+                        // no-op
+                    }
+                }
+            }
+        } catch (Exception e) {
+            String error = "Parsing failed with " + e.getMessage()
+                    + " while accessing " + urlString;
+            Timber.e(error);
+            return new DocumentFetchResult(error, 0);
+        }
+
+        return new DocumentFetchResult(doc, inputStreamResult.isOpenRosaResponse(), inputStreamResult.getHash());
+    }
+
+
+    /**
+     * Creates a http connection and sets up an input stream.
+     *
+     * @param downloadUrl uri of the stream
+     * @param contentType check the returned Mime Type to ensure it matches. "text/xml" causes a Hash to be calculated
+     * @return HttpGetResult - An object containing the Stream, Hash and Headers
+     * @throws Exception - Can throw a multitude of Exceptions, such as MalformedURLException or IOException
+     */
+    public @NonNull
+    HttpGetResult getHttpInputStream(@NonNull String downloadUrl, @Nullable final String contentType) throws Exception {
+        URI uri;
+        try {
+            // assume the downloadUrl is escaped properly
+            URL url = new URL(downloadUrl);
+            uri = url.toURI();
+        } catch (MalformedURLException | URISyntaxException e) {
+            Timber.e(e, "Unable to get a URI for download URL : %s  due to %s : ", downloadUrl, e.getMessage());
+            throw e;
+        }
+
+        if (uri.getHost() == null) {
+            throw new Exception("Invalid server URL (no hostname): " + downloadUrl);
+        }
+
+        return httpInterface.get(uri, contentType, WebCredentialsUtils.getInstance().getCredentials(uri));
+    }
+
+
+    public static String getPlainTextMimeType() {
+        return "text/plain";
+    }
+
+
+    public static class Outcome {
+        public Uri authRequestingServer;
+        public HashMap<String, String> messagesByInstanceId = new HashMap<>();
+    }
+
```

Comment:
Just double check for stray lines after methods and ones after method signatures!